dlio/h5bench_dlio.c

```diff
@@ -0,0 +1,805 @@
+    hid_t    file_id    = H5Fopen(file_path, H5F_ACC_RDONLY, FAPL);
+    hid_t    dataset_id = H5Dopen(file_id, config.RECORDS_DATASET_NAME, DAPL);
+    hid_t    filespace  = H5Dget_space(dataset_id);
+    hid_t    memspace   = H5Screate_simple(3, count, NULL);
+    H5Sselect_hyperslab(filespace, H5S_SELECT_SET, offset, NULL, count, NULL);
+    uint64_t t2 = get_time_usec_return_uint64();
+    assert(file_id >= 0);
+    assert(dataset_id >= 0);
+    assert(filespace >= 0);
+    assert(memspace >= 0);
+
+    uint8_t *data = (uint8_t *)malloc(DIM * DIM * sizeof(uint8_t));
+    if (data == NULL) {
+        exit(1);
+    }
+
+    uint64_t t3     = get_time_usec_return_uint64();
+    herr_t   status = H5Dread(dataset_id, H5T_STD_U8LE, memspace, filespace, DXPL, data);
+    uint64_t t4     = get_time_usec_return_uint64();
+    assert(status >= 0);
+
+    free(data); // TODO: free memory only after compute() call?
+
+    uint64_t t5 = get_time_usec_return_uint64();
+    H5Sclose(memspace);
+    H5Sclose(filespace);
+    H5Dclose(dataset_id);
+    H5Fclose(file_id);
+    uint64_t t6 = get_time_usec_return_uint64();
+
+    *metadata_time_out = (t2 - t1) + (t6 - t5);
+    *read_time_out     = t4 - t3;
+
+    compute(config.PREPROCESS_TIME, config.PREPROCESS_TIME_STDEV);
+}
+
+uint64_t
+compute(float time, float time_stdev)
+{
+    if (time != 0.0 || time_stdev != 0.0) {
+        int t = (uint64_t)(generate_normal_random(time, time_stdev) * 1000000.0);
+        usleep(t > 0 ? t : 0);
+        return t;
+    }
+    return 0;
+}
+
+void
+eval_without_workers(uint32_t epoch, uint32_t *indices, uint64_t *local_metadata_time_out,
+                     uint64_t *local_read_time_out)
+{
+    uint32_t offset = MY_RANK * config.NUM_EVAL_BATCHES_PER_RANK;
+
+    uint64_t t0 = get_time_usec_return_uint64();
+    for (uint32_t i = 0; i < config.NUM_EVAL_BATCHES_PER_RANK; i++) {
+        for (uint32_t j = 0; j < config.BATCH_SIZE_EVAL; j++) {
+            uint32_t file_num =
+                indices[offset + i * config.BATCH_SIZE_EVAL + j] / config.NUM_SAMPLES_PER_FILE + 1;
+            uint32_t sample_num =
+                indices[offset + i * config.BATCH_SIZE_EVAL + j] % config.NUM_SAMPLES_PER_FILE;
+            char file_path[256];
+            snprintf(file_path, sizeof(file_path), "%s/%s/%s_%u_of_%u.h5", config.DATA_FOLDER,
+                     config.VALID_DATA_FOLDER, config.FILE_PREFIX, file_num, config.NUM_FILES_EVAL);
+
+            uint64_t metadata_time = 0, read_time = 0;
+            read_sample(file_path, sample_num, &metadata_time, &read_time);
+
+            *local_metadata_time_out += metadata_time;
+            *local_read_time_out += read_time;
+        }
+
+        batch_loaded_eval(epoch, t0);
+
+        uint64_t t = compute(config.EVAL_TIME, config.EVAL_TIME_STDEV);
+        batch_processed_eval(epoch, t, t0);
+        MPI_Barrier(MPI_COMM_WORLD);
+
+        t0 = get_time_usec_return_uint64();
+    }
+
+    //    TODO: drop_data = False
```

Comment:
If we do not need this code in the near feature, could we remove to have a clean PR?